cpp/src/fil/infer.cu

```diff
@@ -284,32 +329,68 @@ struct tree_aggregator_t<NITEMS, GROVE_PER_CLASS_MANY_CLASSES>
                                                size_t data_row_size)
-    : finalize_block(shared_workspace, num_classes_),
-      per_class_margin(
+    : tmp_storage(shared_workspace),
+      num_classes(num_classes_),
+      per_class_value(
         (vec<NITEMS, float>*)((char*)shared_workspace + data_row_size)) {
     for (int c = threadIdx.x; c < num_classes; c += blockDim.x)
-      per_class_margin[c] = vec<NITEMS, float>();  // initialize to 0.0f
+      per_class_value[c] = vec<NITEMS, float>();  // initialize to 0.0f
     // __syncthreads() is called in infer_k
   }
 
   __device__ __forceinline__ void accumulate(
     vec<NITEMS, float> single_tree_prediction, int tree) {
     // since threads are assigned to consecutive classes, no need for atomics
-    per_class_margin[tree % num_classes] += single_tree_prediction;
+    per_class_value[tree % num_classes] += single_tree_prediction;
     // __syncthreads() is called in infer_k
   }
 
-  __device__ __forceinline__ void finalize(float* out, int num_rows,
-                                           int num_outputs) {
-    // reduce per-class candidate margins to one best class candidate
-    // per thread (for each of the NITEMS rows)
-    vec<NITEMS, best_margin_label> best({-1, -INFINITY});
+  template <typename op>
+  __device__ __forceinline__ vec<NITEMS, float> allreduce_shmem(
+    vec<NITEMS, float>* values, int size) {
+    // assuming size >= blockDim.x, else why the invocation?
+    auto partial = values[threadIdx.x];
+    for (int i = threadIdx.x + blockDim.x; i < size; i += blockDim.x)
+      partial = Vectorized<op>()(partial, values[i]);
+    return block_allreduce<op>(partial, blockDim.x, &values[size]);
+  }
 
+  __device__ __forceinline__ void block_allsoftmax() {
+    // subtract max before exponentiating for numerical stability
+    auto max = allreduce_shmem<cub::Max>(per_class_value, num_classes);
+    for (int c = threadIdx.x; c < num_classes; c += blockDim.x)
+      per_class_value[c] = shifted_exp(per_class_value[c], max);
+    // sum of exponents
+    auto soe = allreduce_shmem<cub::Sum>(per_class_value, num_classes);
+    // softmax phase 2: normalization
     for (int c = threadIdx.x; c < num_classes; c += blockDim.x)
-      best = ArgMax()(best, to_vec(c, per_class_margin[c]));
+      for (int i = 0; i < NITEMS; ++i)
+        per_class_value[c][i] /= soe[i];  // TODO: do bank conflicts matter?
+  }
 
-    __syncthreads();  // free up per_class_margin[]
-    write_best_class_in_block(best, blockDim.x, out, num_rows);
+  __device__ __forceinline__ void finalize(float* out, int num_rows,
+                                           int num_outputs) {
+    if (num_outputs == 1) {  // will output class
+      // reduce per-class candidate margins to one best class candidate
+      // per thread (for each of the NITEMS rows)
+      vec<NITEMS, best_margin_label> best({-1, -INFINITY});
+
+      for (int c = threadIdx.x; c < num_classes; c += blockDim.x)
```

Comment:
this might seem like it fits into reduce_shmem(), but note that:
1. it needs a map before reduce
2. the map changes [increases] the size of the values/array
3. the map depends on the indices
By the time I would be done abstracting that into a more generic reduce_shmem() (or two functions), it would become abstruse.
In a different PR, I can, however, unify these lines with what's happening in finalizing CATEGORICAL_LEAF best class.

Reply:
In this particular case, `reduce_shmem()` functionality is split between the code here (in-thread reduction) and the code in `write_best_class_in_block()` (reduction between threads). If you move the reduction away from `write_best_class_in_block()`, the code in this function would fit nicely into something like `map_reduce_shmem()`. `reduce_shmem()` doesn't have many parameters, so an extra one in `map_reduce_shmem()` won't make the code too complex.